File: Data/Java/union find disjoint set.java
public class UnionFind {

    int[] parent;
    int[] rank;

    // Constructor to initialize sets
    public UnionFind(int size) {
        parent = new int[size];
        rank = new int[size];

        // Each node is its own parent initially
        for (int i = 0; i < size; i++) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    // Find with path compression
    public int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);  // Compress the path
        return parent[x];
    }

    // Union by rank
    public void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        // Merge smaller tree under bigger tree
        if (rootX != rootY) {
            if (rank[rootX] < rank[rootY])
                parent[rootX] = rootY;
            else if (rank[rootX] > rank[rootY])
                parent[rootY] = rootX;
            else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }
    }

    public static void main(String[] args) {
        UnionFind uf = new UnionFind(5);  // Nodes 0-4

        uf.union(0, 1);
        uf.union(1, 2);

        System.out.println(uf.find(0) == uf.find(2));  // true
        System.out.println(uf.find(0) == uf.find(4));  // false
    }
}

File: Data/Python/union find disjoint set.py
class DisjointSet:
    """
    Union-Find (Disjoint Set):
    - Supports union and find operations.
    - Used in Kruskal’s algorithm, detecting cycles.
    - Optimized with path compression and union by rank.
    - Time Complexity: nearly O(1) per operation (amortized)
    """
    def __init__(self, n):
        self.parent = [i for i in range(n)]  # Initially each node is its own parent
        self.rank = [0] * n  # Rank for optimization

    def find(self, x):
        # Path compression: set parent[x] to root directly
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        # Find roots of both sets
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return  # Already in the same set

        # Union by rank: attach smaller tree under larger
        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

# ✅ Example usage
ds = DisjointSet(5)
ds.union(0, 1)
ds.union(1, 2)
print("Find(0):", ds.find(0))  # Output: 0 or root node
print("Find(2):", ds.find(2))  # Output: same as Find(0)
print("Find(3):", ds.find(3))  # Output: 3 (separate set)
